### PR TITLE
Add Restart Mission, hasSeenCompletedMission mutation, bug fixes

### PR DESCRIPTION
**Pull request by amaliwanag2**
bug fixes for tab logging and completing a mission

### CODE REVIEW

File: graphql/config.js [hunk 1]
```diff
@@ -22,4 +22,5 @@ module.exports = {
   SENTRY_PROJECT_ID: process.env.GQL_SENTRY_PROJECT_ID,
   SENTRY_STAGE: process.env.GQL_SENTRY_STAGE,
   SENDGRID_API_KEY: process.env.SENDGRID_API_KEY,
+  E2E_MISSIONS_TEST_TAB_GOAL: process.env.E2E_MISSIONS_TEST_TAB_GOAL,
```
Comment (amaliwanag2) [inline graphql/config.js:25]: @kmjennison can you add this to travis?

File: graphql/database/invitedUsers/utils.js [hunk 2]
```diff
@@ -64,16 +63,10 @@ export const verifyAndSendInvite = async ({
           name: inviterName,
         },
       })
-      await Promise.all([
-        UserModel.update(override, {
-          id: existingUser.id,
-          pendingMissionInvites,
-        }),
-        UserMissionModel.create(override, {
```
Comment (amaliwanag2) [inline graphql/database/invitedUsers/utils.js:72]: we now create this when a user accepts an invite instead

File: graphql/config.js (hunk 1 above)
Comment (kmjennison) [reply graphql/config.js:25]: @amaliwanag2 Add this in .travis.yml

File: graphql/database/missions/hasSeenCompletedMission.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import UserMissionModel from './UserMissionModel'
+import UserModel from '../users/UserModel'
+
+const override = getPermissionsOverride(MISSIONS_OVERRIDE)
+/**
+ * Records a user's acknowledging a mission notification.
+ *
+ * @param {string} userId - The user's ID.
+ * @param {string} missionId - The mission's ID.
+ * @param {string} action - What action the user took (MISSION_COMPLETE, MISSION_STARTED)
```
Comment (kmjennison) [inline graphql/database/missions/hasSeenCompletedMission.js:14]: Update comment

File: graphql/database/missions/completeMission.js [hunk 4]
```diff
@@ -3,41 +3,45 @@ import getCurrentUserMission from './getCurrentUserMission'
...
+    prev.tabStreak.longestTabStreak > current.tabStreak.longestTabStreak
+      ? prev
+      : current
   )
   const mostDayTabsMember = mission.squadMembers.reduce((prev, current) =>
-    prev.missionMaxTabsDay > current.missionMaxTabsDay ? prev : current
+    prev.missionMaxTabsDay.maxDay.numTabs >
+    current.missionMaxTabsDay.maxDay.numTabs
+      ? prev
+      : current
```
Comment (kmjennison) [inline graphql/database/missions/completeMission.js:44]: Please add tests for these.

File: graphql/database/missions/hasSeenCompletedMission.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+const hasSeenCompletedMission = async (userContext, userId, missionId) => {
+  // Get the mission model and modify it.
+  await Promise.all([
+    UserMissionModel.update(override, {
```
Comment (kmjennison) [inline graphql/database/missions/hasSeenCompletedMission.js:19]: Don't need an override, right?

Comment (kmjennison) [reply graphql/database/missions/hasSeenCompletedMission.js:19]: Same goes for most other places we use `UserMissionModel`:
* updateMissionNotification.js
* getPastUserMissions.js
* getCurrentUserMission.js
* createMission.js

Could you change those, too? Better to use built-in authorizers when we can.

File: graphql/database/users/user-utils.js [hunk 6]
```diff
@@ -75,8 +86,9 @@ export const calculateTabStreak = (maxDayObject, tabStreakObject) => {
       .subtract(1, 'days')
       .startOf('day')
   )
-  let newCurrentTabStreak = tabStreakObject.currentTabStreak
-  let newLongestTabStreak = tabStreakObject.longestTabStreak
+  // if this is invoked, then tabstreak should atleast be 1
+  let newCurrentTabStreak = tabStreakObject.currentTabStreak || 1
+  let newLongestTabStreak = tabStreakObject.longestTabStreak || 1
```
Comment (kmjennison) [inline graphql/database/users/user-utils.js:91]: Add tests.

File: graphql/database/users/user-utils.js [hunk 7]
```diff
@@ -54,7 +58,14 @@ export const calculateMaxTabs = (todayTabCount, maxDayObject) => {
...
       date: moment.utc().toISOString(),
-      numTabs: todayTabCount,
+
+      // we have an issue where we calculate the max tabs on a users entire day
+      // but a user can join a mission at any point in the day.  This ensures we only count
+      // tabs on the first day after the user joined the mission for max tabs
+      numTabs:
+        maxPossibleSquadTabs && maxPossibleSquadTabs < todayTabCount
+          ? maxPossibleSquadTabs + 1
+          : todayTabCount,
```
Comment (kmjennison) [inline graphql/database/users/user-utils.js:68]: Add tests.

File: graphql/database/missions/restartMission.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  const newMission = {
+    id: newMissionId,
+    squadName: previousMission.squadName,
+    acceptedSquadMembers: [userId],
+    pendingSquadMembersExisting: otherTeamMembers,
+  }
+  if (E2E_MISSIONS_TEST_TAB_GOAL === 'true') {
+    newMission.tabGoal = 3
+  }
+  // lower tab goal for end to end tests in test environment
```
Comment (kmjennison) [inline graphql/database/missions/restartMission.js:42]: Move comment up

**Review by kmjennison — CHANGES_REQUESTED** [review]

File: graphql/database/missions/completeMission.js (hunk 4 above)
Comment (amaliwanag2) [reply graphql/database/missions/completeMission.js:44]: this one has tests, just had to update the data shape of the tests for it

File: graphql/database/missions/hasSeenCompletedMission.js (hunk 5 above)
Comment (amaliwanag2) [reply graphql/database/missions/hasSeenCompletedMission.js:19]: updated where we get create or update based on UserId.  Ignored where we fetch all or update all documents based on mission id

**Review by kmjennison — APPROVED** [review]
LGTM 🦊

**Pull request merged**